modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderInstance.java

```diff
@@ -0,0 +1,705 @@
+    logger.debug("getCanvasUser({})", userId);
+    String urlString = canvasUrl + "/api/v1/users/sis_user_id:" + userId + "/profile";
+
+    try {
+      URL url = new URL(urlString);
+      logger.debug("Requesting user: {}, using API: {}", userId, url.toString());
+      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+      connection.setRequestMethod("GET");
+      connection.setDoOutput(true);
+      connection.setRequestProperty("Authorization", "Bearer " + canvasToken);
+      connection.setRequestProperty("User-Agent", OC_USERAGENT);
+
+      String json = IOUtils.toString(new BufferedInputStream(connection.getInputStream()));
+      logger.debug(json);
+
+      if (StringUtils.isBlank(json)) {
+        return null;
+      }
+      JsonObject user = new JsonParser().parse(json).getAsJsonObject();
+
+      String canvasEmail = user.get("primary_email").getAsString();
+      String canvasDisplayName = user.get(userNameProperty).getAsString();
+
+      return new String[]{canvasEmail, canvasDisplayName};
+
+    } catch (FileNotFoundException fnf) {
+      logger.debug("User {} does not exist in Canvas: {}", userId, fnf);
+    } catch (Exception e) {
+      logger.warn("Exception getting Canvas user information for user {} at {}: {}", userId, urlString, e);
+    }
+
+    return null;
+  }
+
+  /**
+   * {@inheritDoc}
+   * 
+   * @see org.opencastproject.security.api.CachingUserProviderMXBean#getCacheHitRatio()
+   */
+  @Override
+  public float getCacheHitRatio() {
+    if (requests.get() == 0) {
+      return 0;
+    }
+    return (float) (requests.get() - canvasLoads.get()) / requests.get();
+  }
+
+  /**
+   * Build a Opencast role for course
+   * 
+   * @param courseId
+   * @param canvasRole
+   * @return
+   */
+  private String buildOpencastRole(String courseId, String canvasRole) {
+
+    // Map Canvas role to LTI role
+    String ltiRole = instructorRoles.contains(canvasRole) ? LTI_INSTRUCTOR_ROLE : LTI_LEARNER_ROLE;
+
+    return courseId + "_" + ltiRole;
+  }
+
+  @Override
+  public Iterator<User> findUsers(String query, int offset, int limit) {
+    if (query == null) {
+      logger.error("Query is null");
+      throw new IllegalArgumentException("Query must be set");
+    }
+
+    if (query.endsWith("%")) {
+      query = query.substring(0, query.length() - 1);
+    }
+
+    if (query.isEmpty()) {
+      return Collections.emptyIterator();
+    }
+
+    // Verify if a user exists (non-wildcard searches only)
+    if (!verifyCanvasUser(query)) {
+      return Collections.emptyIterator();
+    }
+
+    List<User> users = new LinkedList<User>();
+    JaxbOrganization jaxbOrganization = JaxbOrganization.fromOrganization(organization);
+    JaxbUser queryUser = new JaxbUser(query, PROVIDER_NAME, jaxbOrganization, new HashSet<JaxbRole>());
+    users.add(queryUser);
+
+    return users.iterator();
+  }
+
+  @Override
+  public Iterator<User> getUsers() {
+    // We never enumerate all users
```

Comment:
Why? That breaks the promise of this method, doesn't it?

Reply:
Yet again, this seems to be copied code from the [moodle provider](https://github.com/opencast/opencast/blob/f5fd263fb90f7eea0c984bf9a63d6cb87c6e297f/modules/userdirectory-moodle/src/main/java/org/opencastproject/userdirectory/moodle/MoodleUserProviderInstance.java#L239)

Perhaps it would be best to leave these methods as a //TODO for the moment?  Elsewise, it might take a little while to get this in an acceptable state, since I don't really have too much time to develop for this.

Reply:
The behaviour for getUsers(), countUsers() and getRoles() is the same across all the providers. No-one has put forward any integration scenarios in which this is required, and it would also not be very efficient to do so.